crates/node/engine/src/attributes.rs

```diff
@@ -0,0 +1,324 @@
+//! Contains a utility method to check if attributes match a block.
+
+use alloy_primitives::{Address, B256};
+use alloy_rpc_types_eth::{Block, Withdrawals};
+use kona_genesis::RollupConfig;
+use kona_rpc::OpAttributesWithParent;
+use op_alloy_rpc_types::Transaction;
+
+/// Represents whether the attributes match the block or not.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum AttributesMatch {
+    /// The attributes match the block.
+    Match,
+    /// The attributes do not match the block.
+    Mismatch(AttributesMismatch),
+}
+
+impl AttributesMatch {
+    /// Returns true if the attributes match the block.
+    pub const fn is_match(&self) -> bool {
+        matches!(self, Self::Match)
+    }
+
+    /// Returns true if the attributes do not match the block.
+    pub const fn is_mismatch(&self) -> bool {
+        matches!(self, Self::Mismatch(_))
+    }
+
+    /// Checks that withdrawals for a block and attributes match.
+    pub fn check_withdrawals(
+        config: &RollupConfig,
+        attributes: &OpAttributesWithParent,
+        block: &Block<Transaction>,
+    ) -> Self {
+        let attr_withdrawals = attributes.attributes.payload_attributes.withdrawals.as_ref();
+        let attr_withdrawals = attr_withdrawals.map(|w| Withdrawals::new(w.to_vec()));
+        let block_withdrawals = block.withdrawals.as_ref();
+
+        if config.is_canyon_active(block.header.timestamp) {
+            // In canyon, the withdrawals list should be some and empty
+            if attr_withdrawals.is_none_or(|w| !w.is_empty()) {
+                return Self::Mismatch(AttributesMismatch::CanyonWithdrawalsNotEmpty);
+            }
+            if block_withdrawals.is_none_or(|w| !w.is_empty()) {
+                return Self::Mismatch(AttributesMismatch::CanyonWithdrawalsNotEmpty);
+            }
+            if !config.is_isthmus_active(block.header.timestamp) {
+                // In canyon, the withdrawals root should be set to the empty value
+                let empty_hash = alloy_consensus::EMPTY_ROOT_HASH;
+                if block.header.inner.withdrawals_root != Some(empty_hash) {
+                    return Self::Mismatch(AttributesMismatch::CanyonNotEmptyHash);
+                }
+            }
+        } else {
+            // In bedrock, the withdrawals list should be None
+            if attr_withdrawals.is_some() {
+                return Self::Mismatch(AttributesMismatch::BedrockWithdrawals);
+            }
+        }
+
+        if config.is_isthmus_active(block.header.timestamp) {
+            // In isthmus, the withdrawals root must be set
+            if block.header.inner.withdrawals_root.is_none() {
```

Comment:
I think this is fine; Properly checking the WD root against the state root would take a while and this is a particularly hot path.. Does the `op-node` also not actually check this?

Reply:
I mirrored what the `op-node` did here yes.